docs/Toy/toy/rewrites/inline_toy.py

```diff
@@ -0,0 +1,102 @@
+from xdsl.ir import MLContext
+from xdsl.dialects.builtin import (
+    ModuleOp,
+    StringAttr,
+)
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    PatternRewriteWalker,
+    op_type_rewrite_pattern,
+    RewritePattern,
+    PatternRewriter,
+)
+from xdsl.transforms.dead_code_elimination import dce
+
+from ..dialects import toy
+
+
+class InlineFunctions(RewritePattern):
+    _func_op_by_name: dict[str, toy.FuncOp] | None = None
+
+    def func_op(self, module: ModuleOp, name: str) -> toy.FuncOp:
+        if self._func_op_by_name is None:
+            self._func_op_by_name = {
+                op.sym_name.data: op for op in module.ops if isinstance(op, toy.FuncOp)
+            }
+        return self._func_op_by_name[name]
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: toy.GenericCallOp, rewriter: PatternRewriter):
+        """
+        For each generic call, find the function that it calls, and inline it.
+        """
+
+        # Get module
+        parent = op.parent_op()
+        assert isinstance(parent, toy.FuncOp)
+        module = parent.parent_op()
+        assert isinstance(module, ModuleOp)
+
+        # Clone called function
+        impl = self.func_op(module, op.callee.string_value()).clone()
+        impl_block = impl.body.block
+
+        # Cast operands to unranked
+        inputs = [toy.CastOp(operand) for operand in op.operands]
```

Comment:
I don't understand why you need that, but I trust you on this!

Reply:
This is actually a requirement of the inliner interface in MLIR, we don't necessarily need it it Toy, yet, but I thought I'd make the IR look exactly the same as the MLIR tutorial if possible